distribution/push.go

```diff
@@ -158,13 +160,14 @@ func Push(ctx context.Context, ref reference.Named, imagePushConfig *ImagePushCo
 // is finished. This allows the caller to make sure the goroutine finishes
 // before it releases any resources connected with the reader that was
 // passed in.
-func compress(in io.Reader) (io.ReadCloser, chan struct{}) {
+func compress(in io.Reader, compressionThreads int) (io.ReadCloser, chan struct{}) {
 	compressionDone := make(chan struct{})
 
 	pipeReader, pipeWriter := io.Pipe()
 	// Use a bufio.Writer to avoid excessive chunking in HTTP request.
 	bufWriter := bufio.NewWriterSize(pipeWriter, compressionBufSize)
-	compressor := gzip.NewWriter(bufWriter)
+
+	compressor := newCompressor(bufWriter, compressionThreads)
```

Comment:
pgzip has its own internal buffering, so using the `bufio.Writer` is just allocating/copying without any benefit. Consider adding buffering to internal gzip only in `newCompressor`.